CLI Samples launcher + Vulkan Hello Triangle

TomAtkinsonArm opened the pull request:
## Description

Blocked by #485 #487 #488 #490

Large diff as this contains commits from several branches

Currently only works on Ubuntu with XLIB enabled as I have not added every window and surface type yet

```
loading sample configs
candidate: /build/linux/cmd/samples_launcher/samples.json
selected: /build/linux/cmd/samples_launcher/samples.json
looking for: libsample__hello_sample.so - not found
looking for: libsample__hello_triangle.so
candidate: ./build/linux/samples/hello_triangle/lib/Debug/x86_64/libsample__hello_triangle.so
selecting: ./build/linux/samples/hello_triangle/lib/Debug/x86_64/libsample__hello_triangle.so


Available Samples: (1/2)

id:             hello_triangle
name:           Hello Triangle
description:    Vulkan Hello Triangle
compile target: sample__hello_triangle

Hello Triangle!
```

`samples.json` is maintained by CMake automatically and contains metadata for all registered samples. Samples may be available or not, we cant easily track which have been built previously and which haven't. The Sample Launcher deduces available samples by searching the current working directory for platform specific runtime libraries for a given sample. Available samples are then displayed in a list.

The code sample shows the launcher being run with `hello_triangle`. The runtime library is found as stated in the lookup log. The launcher then loads this sample and runs its `sample_main`.

![Screenshot from 2022-07-09 18-44-13](https://user-images.githubusercontent.com/48598630/178117153-99bd3cd8-0062-4295-bc76-8e7cf4db2838.png)

![Screenshot from 2022-07-09 18-41-06](https://user-images.githubusercontent.com/48598630/178116953-791e795b-b934-4c49-9fd6-0212c50e7eaa.png)

After a little bit more work I ported Hello Triangle

![Screenshot from 2022-07-09 22-51-06](https://user-images.githubusercontent.com/48598630/178123806-16f530ac-cc74-4217-8c67-4b8420f0e6b5.png)

# Solved Problems

### Compilation

Single samples can be compiled by using the `sample__<sample_name>` target. All samples can be compiled with `vkb__samples`. `samples_launcher` is only compiled when one or more samples are selected for compilation. A user should never need to manually compile `samples_launcher`.

As the framework is now split into components a sample and the launcher will only compile the targets needed. For instance, compiling the current samples yields sub second compile times. There is also no link step so serialized link times in our current builds will not occur in this version of the framework.

### Samples

Samples no longer need to follow strict patterns for creation. Requirements for a sample are:

- A file defining `sample_main`
- `vkb__register_sample` - sample build target
- `vkb__register_sample_descriptor` - sample metadata

From here samples can come in any shape or size. The most appropriate execution flow can be used for each sample. We may want to define a component which acts as a base sample to reduce boilerplate, but these building blocks do not need to be the same for all samples.

# TODO

- [ ] - condense launcher code
- [ ] - add a default samples build test which makes sure the launcher can find all samples when building with `vkb__samples`

TomAtkinsonArm committed:
event bus

TomAtkinsonArm committed:
add AbstractChannel comment

TomAtkinsonArm committed:
add pipelines

TomAtkinsonArm committed:
stages with custom fields

TomAtkinsonArm committed:
CUSTOM_MAIN

TomAtkinsonArm committed:
CUSTOM_MAIN

TomAtkinsonArm committed:
test for correct context

TomAtkinsonArm committed:
CMake WIN32

TomAtkinsonArm committed:
dynamic samples

TomAtkinsonArm committed:
add more comments

TomAtkinsonArm committed:
use window.h

TomAtkinsonArm committed:
windows fix

TomAtkinsonArm committed:
Pipeline -> EventPipeline

TomAtkinsonArm committed:
fix macos lib name

TomAtkinsonArm committed:
window interface

TomAtkinsonArm committed:
headless window

TomAtkinsonArm committed:
GLFW window

TomAtkinsonArm committed:
GLFW window

TomAtkinsonArm committed:
add VKB_TEST_GLFW

TomAtkinsonArm committed:
add VKB_TEST_GLFW

TomAtkinsonArm committed:
add VKB_TEST_GLFW

TomAtkinsonArm committed:
fix glfw errors

TomAtkinsonArm committed:
main

TomAtkinsonArm committed:
encoding

TomAtkinsonArm committed:
launcher base

TomAtkinsonArm committed:
fix test

TomAtkinsonArm committed:
config

TomAtkinsonArm committed:
add samples

TomAtkinsonArm committed:
ci fixed

TomAtkinsonArm committed:
rebase fixes

TomAtkinsonArm committed:
deduce available samples

TomAtkinsonArm committed:
hello triangle

TomAtkinsonArm committed:
add sleep to prevent resource drain

TomAtkinsonArm committed:
windows build fixes

TomAtkinsonArm committed:
std filesystem

TomAtkinsonArm committed:
partial fs fixes

TomAtkinsonArm committed:
windows builds and runs hello_triangle

TomAtkinsonArm committed:
add samples_launcher window improvements

TomAtkinsonArm committed:
fix unused param warnings

TomAtkinsonArm committed:
Event Bus (#485)

* event bus

* add AbstractChannel comment

* response to feedback

* add assert

* assert

* added observer expiration test

gpx1000 commented on bldsys/cmake/component_helpers.cmake:
```diff
@@ -32,6 +32,8 @@ function(vkb__register_component)
 
         add_library("vkb__${TARGET_NAME}" STATIC ${TARGET_SRC})
 
+        set_property(TARGET "vkb__${TARGET_NAME}" PROPERTY POSITION_INDEPENDENT_CODE ON) # PIC required for shared lib linking
```
it is probably better to use set_compile_options("vkb_${TARGET_NAME}" PRIVATE POSITION_INDEPENDENT_CODE ON) 
NB: prior to CMake 3.19, -fPIC didn't mean that pie is also by default mandated in cmake other than in Android builds.  I think we're mandating a recent enough version for this to not be an issue.  Just listing here as a just incase.

gpx1000 commented on bldsys/cmake/sample_helper.cmake:
```diff
@@ -17,159 +17,57 @@
+        endif()
+        
+        if(MSVC)
+            target_compile_options("sample__${TARGET_NAME}" PRIVATE /W4 /WX)
```
Dunno if you get any warnings in MSVC when compiling with this due to /W3 and /Md being set automatically D9025 might crop up.  However, just in case it happens:
https://cmake.org/cmake/help/v3.15/policy/CMP0092.html
https://cmake.org/cmake/help/v3.15/policy/CMP0091.html
add:
if (POLICY CMP0091)
  cmake_policy(SET CMP0091 NEW)
endif ()
if (POLICY CMP0092)
  cmake_policy(SET CMP0092 NEW)
endif ()

TomAtkinsonArm committed:
Visual Studio Folders (#496)

* fix folder structure in Visual Studio

* add interface library support

* review changes

TomAtkinsonArm committed:
STD Filesystem VFS (#495)

* std filesystem

more tests and fixes

recursive tests

more tests

remove order specific checks from tests

review comments

* fix duplicated check

* review changes

TomAtkinsonArm committed:
Event Pipelines (#487)

* event bus

* response to feedback

* add assert

* assert

* added observer expiration test

* add pipelines

* stages with custom fields

* Pipeline -> EventPipeline

* add comments

TomAtkinsonArm committed:
directly create a surface

gpx1000 commented on bldsys/cmake/component_helpers.cmake:
```diff
@@ -16,21 +16,37 @@
         message("ADDING STATIC: vkb__${TARGET_NAME}")
 
-        add_library("vkb__${TARGET_NAME}" STATIC ${TARGET_SRC})
+        add_library("vkb__${TARGET_NAME}" STATIC ${TARGET_SRC} ${TARGET_HEADERS})
```
think this is the same as in 490?  Not sure of the advantage of adding headers explicitly to the cmake target.

gpx1000 commented on bldsys/cmake/sample_helper.cmake:
```diff
@@ -17,159 +17,57 @@
-        SHADER_FILES_GLSL
-            ${TARGET_SHADER_FILES_GLSL})
+    get_property(DESCRIPTORS GLOBAL PROPERTY SAMPLE_DESCRIPTORS)
+    list(APPEND DESCRIPTORS "{\"id\":\"${TARGET_ID}\",\"name\":\"${TARGET_NAME}\", \"description\": \"${TARGET_DESCRIPTION}\", \"library_name\": \"${TARGET_LIB}\"}")
```
could potentially use jq or https://github.com/sbellus/json-cmake or some such to make this a usable thing at the cmake level. however, this works.

TomAtkinsonArm replied:
Does this allow for creating json structs in cmake? Or just parsing?

TomAtkinsonArm commented:
@SaschaWillems @gary-sweet @asuessenbach can you please take a look at this PR when you get a chance? Looking for general feedback on how these components fit together and how samples can be written. The idea is to remove all the constraints on samples so that they can be more flexible in the future.

An example of this is `sample_main` which is all that is required to create a sample. Or `EventPipeline` which allows a sample to define its own lifecycle. "Frameworks" can be extracted into a series of default configurations for these components as most components are configurable or extendable

asuessenbach commented:
First of all: the framework compiles without any issues with VS2022.
But when I start the `samples_launcher`, I get the following message:
```
loading sample configs
[vfs/std_filesystem.cpp:208] path does not exist: C:\git\Vulkan-Samples\build\VS2022_64\cmd\samples_launcher\build
```
And, indeed, that path does not exist. I found the file `samples.json` directly in the `samples_launcher` directory; there's no build directory there.
For a thorough review, I might need some more time. After all, this is pretty big.

asuessenbach commented on cmd/samples_launcher/include/config.hpp:
```diff
@@ -0,0 +1,42 @@
+};
+
+// Load a config from a binary buffer
+components::StackErrorPtr load_config_from_json(const std::vector<uint8_t> &data, Config *config);
```
I think, we agreed to switch to `std::exception`s here, instead of `components::StackErrorPtr`?
(Funnily, `components::StackErrorPtr` derives from `std::exception`, but is not used by the `throw`/`catch` mechanism of exceptions!)

asuessenbach commented on cmd/samples_launcher/src/main.cpp:
```diff
@@ -0,0 +1,173 @@
+
+	std::cout << "loading sample configs\n";
+
+	std::vector<std::string> files;
```
Maybe rename `files` to `config_files`, or so?

asuessenbach commented on cmd/samples_launcher/src/main.cpp:
```diff
@@ -0,0 +1,173 @@
+
+		std::cout << "looking for: " << library_name;
+
+		std::vector<std::string> candidates;
```
Do you expect more than one candidate per sample? But you just use the first one.
Maybe the search in the sample_files could then be done by `std::find_if`:
```
		auto sample_it = std::find_if(sample_files.begin(),
		                              sample_files.end(),
		                              [&library_name](auto const &sample_file) { return vfs::helpers::get_file_name(sample_file) == library_name; });
		if (sample_it == sample_files.end())
		{
			std::cout << " - not found\n";
			continue;
		}

```

asuessenbach commented on cmd/samples_launcher/src/main.cpp:
```diff
@@ -0,0 +1,173 @@
+		return EXIT_SUCCESS;
+	}
+
+	for (auto &info : available_samples)
```
Maybe use `std::find_if`, again?
```
  auto sample_it =
	    std::find_if(available_samples.begin(), available_samples.end(), [&args](auto &sample_info) { return sample_info.sample.id == args[0]; });
	if (sample_it == available_samples.end())
	{
		std::cout << "Missing sample " << args[0] << "\n";
	}

```

asuessenbach commented on components/encoding/include/components/encoding/json.hpp:
```diff
@@ -0,0 +1,130 @@
+
+#define HANDLE_JSON_ERROR_START() \
+	try                           \
+	{
```
I would not hide scope opening in a macro. Functions like marshall_json below look somewhat awkward then.
Maybe just some HANDLE_JSON_ERROR representing the catch block would be enough?

asuessenbach commented on cmd/samples_launcher/include/config_marshalers.hpp:
```diff
@@ -0,0 +1,162 @@
+}
+
+template <>
+class JsonMarshaler<config::Sample>
```
The termini `Marshaler` and `UnMarshaler` are a bit alien to me.
Maybe something like `Serializer`/`Deserializer` or `Encoder`/`Decoder` would be a bit more descriptive?

asuessenbach commented on components/events/src/input_manager.cpp:
```diff
@@ -59,18 +66,18 @@ void InputManager::process(ChannelReceiverPtr<TouchEvent> &events)
 		{
-			touch.position.x = event.pos_x - touch.position.x;
-			touch.position.y = event.pos_y - touch.position.y;
+			touch.position.x = event.x - touch.position.x;
```
What is this supposed to do?
It's unconditionally overwritten next.

asuessenbach commented on components/events/tests/event_bus.test.cpp:
```diff
@@ -0,0 +1,348 @@
+
+	inline size_t unobserved_each_event_count() const
+	{
+		size_t unprocessed_events{0};
```
Maybe
```
  return std::accumulate(
    m_each_callbacks.begin(), m_each_callbacks.end(), 0, []( size_t acc, auto const & callback ) { return acc + callback.second->queue_size(); } );
```
And similar for the other count functions.

asuessenbach commented on components/events/tests/event_pipeline.test.cpp:
```diff
@@ -0,0 +1,313 @@
+	// on a second pass we expect only the "then" stages to be processed
+	execution_index = 2;
+
+	pipeline.process();
```
I don't see how the correct order is verified. Could you please elaborate a bit?

asuessenbach commented on components/events/tests/event_pipeline.test.cpp:
```diff
@@ -0,0 +1,313 @@
+
+	REQUIRE(one_execution_count == 1);
+	REQUIRE(two_execution_count == 4);
+	REQUIRE(three_execution_count == 4);
```
I see, the number of calls of each event is checked, but how is the correct order verified?

asuessenbach commented on components/platform/include/components/platform/dl.hpp:
```diff
@@ -0,0 +1,81 @@
+#	undef IMPORT
+#	define EXPORT __declspec(dllexport)
+#	define IMPORT __declspec(dllimport)
+#endif
```
Maybe better to read:
```
#if defined(_WIN32)
#  define EXPORT __declspec(dllexport)
#  define IMPORT __declspec(dllimport)
#else
#  define EXPORT
#  define IMPORT
#endif
```

asuessenbach commented on components/platform/include/components/platform/dl.hpp:
```diff
@@ -0,0 +1,81 @@
+ * @param name library name
+ * @return std::string OS library name
+ */
+std::string os_library_name(const std::string &name);
```
Mixing functions taking a `const std::string &` with functions taking a `const char *` in the same header seems counter-intuitive. I would stay with one of them, keeping the same level of abstraction.

asuessenbach commented on components/platform/include/components/platform/platform.hpp:
```diff
@@ -0,0 +1,196 @@
+	 */
+	struct PlatformContext
+	{
+		virtual std::vector<std::string> arguments() const = 0;
```
Why do you make this function pure virtual?
Probably all Context types just hold the arguments in a vector, that is it could be as well be moved into `PlatformContext`. Then, you don't even need that accessor function.

asuessenbach commented on components/platform/include/components/platform/platform.hpp:
```diff
@@ -0,0 +1,196 @@
+	 *
+	 * @return int status code
+	 */
+	int platform_main(components::PlatformContext *);
```
Shouldn't `platform_main` take a `const &`, or maybe a `const *`?

asuessenbach commented on components/platform/src/dl.cpp:
```diff
@@ -0,0 +1,93 @@
+	std::stringstream lib_name;
+	lib_name << os_library_prefix() << name << os_library_postfix();
+	return lib_name.str();
+}
```
No need to use a `std::stringstream` here. Simply
`	return os_library_prefix() + name + os_library_postfix();
`

asuessenbach commented on components/vfs/src/root_file_system.cpp:
```diff
@@ -256,19 +274,12 @@ StackErrorPtr RootFileSystem::enumerate_folders_recursive(const std::string &fil
 	}
 
-	*folders = std::vector<std::string>{all_dirs.begin(), all_dirs.end()};
+	std::vector<std::string> all_folders{all_dirs.begin(), all_dirs.end()};
```
Why do you copy `all_dirs` into `all_folders` here?

asuessenbach commented on components/vfs/src/std_filesystem.cpp:
```diff
@@ -0,0 +1,256 @@
+
+bool StdFSFileSystem::folder_exists(const std::string &file_path)
+{
+	return std::filesystem::is_directory(m_base_path += file_path);
```
You probably wanted to use `+`, instead of `+=` here?
Then, the function could be `const`.

asuessenbach commented on components/vfs/src/std_filesystem.cpp:
```diff
@@ -0,0 +1,256 @@
+
+bool StdFSFileSystem::file_exists(const std::string &file_path)
+{
+	return std::filesystem::is_regular_file(m_base_path += file_path);
```
As above: `+`, instead of `+=`; and `const`.

asuessenbach commented on components/vfs/src/std_filesystem.cpp:
```diff
@@ -0,0 +1,256 @@
+	return std::filesystem::is_regular_file(m_base_path += file_path);
+}
+
+StackErrorPtr StdFSFileSystem::read_chunk(const std::string &file_path, const size_t offset, const size_t count, std::shared_ptr<Blob> *blob)
```
Could this function be const?

asuessenbach commented on components/vfs/src/std_filesystem.cpp:
```diff
@@ -0,0 +1,256 @@
+	return nullptr;
+}
+
+size_t StdFSFileSystem::file_size(const std::string &file_path)
```
Could be const?

asuessenbach commented on components/vfs/src/std_filesystem.cpp:
```diff
@@ -0,0 +1,256 @@
+	return static_cast<size_t>(std::filesystem::file_size(full_path));
+}
+
+StackErrorPtr StdFSFileSystem::write_file(const std::string &file_path, const void *data, size_t size)
```
Could be const?

asuessenbach commented on components/vfs/src/std_filesystem.cpp:
```diff
@@ -0,0 +1,256 @@
+	return nullptr;
+}
+
+StackErrorPtr StdFSFileSystem::enumerate_files(const std::string &dir, std::vector<std::string> *files)
```
Could be const?

asuessenbach commented on components/vfs/src/std_filesystem.cpp:
```diff
@@ -0,0 +1,256 @@
+	return nullptr;
+}
+
+StackErrorPtr StdFSFileSystem::enumerate_folders(const std::string &dir, std::vector<std::string> *folders)
```
Could be const?

asuessenbach commented on components/windows/src/glfw.cpp:
```diff
@@ -0,0 +1,479 @@
+{
+	// TODO: GLFW_X11_XCB_VULKAN_SURFACE
+
+	if (!glfwInit())
```
Is `glfwTerminate` called somewhere?

asuessenbach commented on components/windows/src/glfw.cpp:
```diff
@@ -0,0 +1,479 @@
+{
+	glfwSetWindowSize(m_handle, (int) extent.width, (int) extent.height);
+	m_extent = extent;
+	if (m_content_rect_sender)
```
Are the checks against the sender needed?
Or should that just be an assertion?

asuessenbach commented on samples/hello_triangle/src/context.cpp:
```diff
@@ -0,0 +1,72 @@
+	if (context.swapchain != VK_NULL_HANDLE)
+	{
+		vkDestroySwapchainKHR(context.device, context.swapchain, nullptr);
+		context.swapchain = VK_NULL_HANDLE;
```
You should either reset all teared-down members here, or none.

asuessenbach commented on samples/hello_triangle/src/device.cpp:
```diff
@@ -0,0 +1,103 @@
+
+	if (!validate_extensions(required_device_extensions, device_extensions))
+	{
+		throw std::runtime_error("Required device extensions are missing, will try without.");
```
What do you mean with "will try without" ?

asuessenbach commented on samples/hello_triangle/src/instance.cpp:
```diff
@@ -0,0 +1,199 @@
+
+bool validate_extensions(const std::vector<const char *> &         required,
+                         const std::vector<VkExtensionProperties> &available)
+{
```
Maybe:
```
	// inner find_if returns true if the extension was not found
	// outer find_if returns true if none of the extensions were not found, that is if all extensions were found
	return std::find_if(required.begin(),
	                    required.end(),
	                    [&available](auto required_extension) {
		                    return std::find_if(available.begin(),
		                                        available.end(),
		                                        [&required_extension](auto const &available_extension) {
			                                        return strcmp(available_extension.extensionName, required_extension) == 0;
		                                        }) == available.end();
	                    }) == required.end();

```

asuessenbach commented on samples/hello_triangle/src/instance.cpp:
```diff
@@ -0,0 +1,199 @@
+bool validate_layers(const std::vector<const char *> &     required,
+                     const std::vector<VkLayerProperties> &available)
+{
+	for (auto extension : required)
```
Maybe:
```
	// inner find_if returns true if the layer was not found
	// outer find_if returns true if none of the layers was not found, that is if all layers were found
	return std::find_if(required.begin(),
	                    required.end(),
	                    [&available](auto required_layer) {
		                    return std::find_if(available.begin(),
		                                        available.end(),
		                                        [&required_layer](auto const &available_layer) {
			                                        return strcmp(available_layer.layerName, required_layer) == 0;
		                                        }) == available.end();
	                    }) == required.end();

```

asuessenbach commented on samples/hello_triangle/src/instance.cpp:
```diff
@@ -0,0 +1,199 @@
+{
+	LOGI("Initializing vulkan instance.");
+
+	if (volkInitialize())
```
Maybe:
`if (volkInitialize() != VK_SUCCESS)`

asuessenbach commented on samples/hello_triangle/src/swapchain.cpp:
```diff
@@ -0,0 +1,390 @@
+		}
+
+		format.format = VK_FORMAT_UNDEFINED;
+		for (auto &candidate : formats)
```
Some `std::find_if` here?

asuessenbach reviewed (commented):
On Windows using VS, I had to set the Working Directory to the root of the repository in order to find everything. I don't see that mentioned anywhere.
